File: java/Text_Justification.java
// Very interesting!
// Using a len and count to record the current length and word counts, when len + new > maxWidth, we find a line answer, and output it.
// Time is O(n), Space is O(n).

public class Solution {
    public List<String> fullJustify(String[] words, int maxWidth) {
        List<String> result = new ArrayList<String>();
        int len = 0, count = 0, start = 0;
        for (int i = 0; i < words.length; ++i) {
            if (len + words[i].length() + count > maxWidth) {
                StringBuilder sb = new StringBuilder();
                int blanks = maxWidth - len;
                if (count == 1) {
                    sb.append(words[start]);
                    for (int j = 0; j < blanks; ++j)
                        sb.append(' ');  
                }
                else {
                    int average = blanks / (count-1);
                    int more = blanks - average * (count-1); 
                    for (int j = 0; j < count-1; ++j) {
                        sb.append(words[start+j]);
                        for (int w = 0; w < average; ++w)
                            sb.append(' ');
                        if (j < more)
                            sb.append(' ');
                    }
                    sb.append(words[start+count-1]);
                }
                result.add(sb.toString());
                start = i;
                len = words[i].length();
                count = 1;   
                continue;       
            }
            len += words[i].length();
            count++;
        }
        if (count > 0) {
            StringBuilder sb = new StringBuilder();
            sb.append(words[start]);
            len = words[start].length();
            for (int i = 1; i < count; ++i) {
                sb.append(" " + words[start+i]);
                len += 1 + words[start+i].length(); 
            }
            for (int i = 0; i < maxWidth - len; ++i)
                sb.append(' ');
            result.add(sb.toString());
        }
        return result;
    }
}

File: Python/Text Justification.py
class Solution:
    # @param {string[]} words
    # @param {integer} maxWidth
    # @return {string[]}
    def fullJustify(self, words, maxWidth):
        LineS = []
        line = []
        Text = []
        length = maxWidth
        wordnum = 0
        for word in words:
            if length >= (wordnum + len(word)):
                line.append(word)
                wordnum += 1
                length -= len(word)
            else:
                line.append(length)
                LineS.append(line)
                length = maxWidth
                wordnum = 1
                line = [word]
                length -= len(word)
        if len(line):
            line.append(length)
            LineS.append(line)

        for i in range( 0, len(LineS) - 1):
            line = LineS[i]
            numspace = line[len(line) - 1]
            wordnum = len(line) - 1
            string = ''
            if wordnum == 1:
                string = line[0] + ' ' * numspace
                Text.append(string)
            else:
                averagespace = numspace // (wordnum-1)
                leftspace = numspace % (wordnum-1)
                for i in range(0, wordnum - 1):
                    if i < leftspace:
                        string += line[i] + ' '*(averagespace + 1)
                    else:
                        string += line[i] + ' '*averagespace
                string += line[wordnum - 1]
                Text.append(string)
        string = ''
        for i in range(0, len(LineS[len(LineS) - 1]) - 1):
            string = string + LineS[len(LineS) - 1][i] + ' '
        string = string[:-1]
        string += ' ' * (maxWidth - len(string))
        Text.append(string)
        return Text
